composition-js/src/merging/merge.ts

```diff
@@ -1102,40 +1106,55 @@ class Merger {
   }
 
   private addArgumentsShallow<T extends FieldDefinition<any> | DirectiveDefinition>(sources: (T | undefined)[], dest: T) {
+    const argNames = new Set<string>();
     for (const source of sources) {
       if (!source) {
         continue;
       }
-      for (const argument of source.arguments()) {
-        if (!dest.argument(argument.name)) {
-          dest.addArgument(argument.name);
-        }
-      }
-    }
-  }
-
-  private mergeArgument(sources: (ArgumentDefinition<any> | undefined)[], dest: ArgumentDefinition<any>, useIntersection: boolean = false) {
-    if (useIntersection) {
-      for (const source of sources) {
-        if (!source) {
+      source.arguments().forEach((arg) => argNames.add(arg.name));
+    }
+
+    for (const argName of argNames) {
+      // We add the arguement unconditionally even if we're going to remove it in
+      // some path. Done because this helps reusing our "reportMismatchHint" method
+      // in those cases.
+      const arg = dest.addArgument(argName);
+      // If all the sources that have the field have the argument, we do merge it
+      // and we're good, but otherwise ...
+      if (!sources.every((s) => !s || s.argument(argName))) {
```

Comment:
Is checking the name of the argument sufficent? Don't we also need to check the type or is that getting checked somewhere else?

Reply:
The type is checked later when we properly call `mergeArgument` (which may error out composition in its own right).